### PR TITLE
176774633 handling constant terms

### PR DESCRIPTION
**Pull request by mstechly**
While working on this with @alexjuda and @max-radin we've uncovered multiple other issues with the code and how it's tested that we'll address in the future.

List of most notable changes:
- constant term is now included in calculating expectation values
- we removed `estimate_nmeas_for_operator` as it could lead to improper behaviour, as the ordering of the terms could easily change during grouping and there's no way to make sure that expectation values are in the right order.
- we removed `hamiltonian_analysis` step as it uses `estimate_nmeas_for_operator` and doesn't seem to be important (this might be have been a bad decision, I specifically ask for your approval @jgonthier)!

Uncovered issues that should be addressed separately:
- Improve tests to include simple Hamiltonians to make them more reliable and make sure they cover all the important cases.
- Add testing specific properties, e.g.: "adding a constant term to hamiltonian doesn't change K2 for estimate_nmeas_for_frames"
- Introduce a mechanism that will ensure that expectation values are in the same order as groups.
- Perhaps refactor `estimate_nmeas_for_frames` so it doesn't return several objects.

### CODE REVIEW

**Review by aakunitsa — COMMENTED** [review]
Looks good, but I think we may want to rewrite `hamiltonian_analysis` to perform grouping and then call `grouped_hamiltonian_analysis` rather than just removing the step altogether. :)

File: src/python/zquantum/core/hamiltonian.py [hunk 1]
```diff
@@ -200,49 +216,18 @@ def get_expectation_values_from_rdms(
...
         raise RuntimeWarning(
             f"Expectation values extracted from rdms inside get_expectation_values_from_rdms are complex!"
         )
     expectations = np.real(expectations)
     np.clip(expectations, -1, 1, out=expectations)
-
     return ExpectationValues(expectations)
 
 
-def estimate_nmeas_for_operator(
```
[the remark this answers is not in the document]
Comment (max-radin) [reply src/python/zquantum/core/hamiltonian.py:222]: It won't solve the problem, but it gets rid of a pathway that's very prone to user error.

Since the function has `ExpectationValues` and `QubitOperator` as parameters, one might naturally expect that the `ExpectationValues` should be in the same order as the corresponding terms of the `QubitOperator`. But this will actually give the wrong result because the expectation values need to be ordered according to how the terms are ordered after grouping.

@mstechly and I discovered that in fact some of the unit tests were making this mistake, but they just happened to edge cases where grouping didn't change the order!

File: steps/measure.py [hunk 2]
```diff
@@ -157,28 +157,6 @@ def evaluate_ansatz_based_cost_function(
     save_value_estimate(value_estimate, "value_estimate.json")
 
 
-def hamiltonian_analysis(
```
[the remark this answers is not in the document]
Comment (max-radin) [reply steps/measure.py:160]: Yes for example you can do:
1. Use the [`group_comeasurable_terms_greedy`](https://github.com/zapatacomputing/z-quantum-core/blob/6f034fa0353e2307cc2e0cfe55e8d343d777b51d/steps/operators.py#L166) step to generate a list of qubit operators.
2. Use the [expectation_values_from_rdms_for_qubitoperator_list](https://github.com/zapatacomputing/z-quantum-core/blob/6f034fa0353e2307cc2e0cfe55e8d343d777b51d/steps/measure.py#L239) step to get the expectation values for the grouped Hamiltonian.
3. Use the [`grouped_hamiltonian_analysis`](https://github.com/zapatacomputing/z-quantum-core/blob/6f034fa0353e2307cc2e0cfe55e8d343d777b51d/steps/measure.py#L182) step to get the measurement estimate.

But I think there's a case to be made that we might want to keep the `hamiltonian_analysis` step for the case where you are using upper bounds on variances instead of expectation values. @aakunitsa would that be helpful for you?

[the remark this answers is not in the document]
Comment (mstechly) [reply steps/measure.py:160]: @max-radin But you can just don't pass any expectation values and it will by default use upper bounds, right?

[the remark this answers is not in the document]
Comment (max-radin) [reply steps/measure.py:160]: If you don't provide any expectation values, `hamiltonian_analysis` should give you the upper bound estimate because of this: https://github.com/zapatacomputing/z-quantum-core/blob/31dc4c22d497ee77c39f0229ed53a6785306daaf/src/python/zquantum/core/hamiltonian.py#L141-L145

File: src/python/zquantum/core/hamiltonian.py [hunk 3]
```diff
@@ -102,6 +109,18 @@ def get_decomposition_function(
     return decomposition_function
 
 
+def _calculate_variance_for_group(group: QubitOperator) -> float:
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/core/hamiltonian.py]: Good point, did that in: https://github.com/zapatacomputing/z-quantum-core/pull/268/commits/34acb63fb579a552a11c1d2f9e5d40deec979a6b

File: steps/measure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstechly) [reply steps/measure.py:160]: > But I think there's a case to be made that we might want to keep the hamiltonian_analysis step for the case where you are using upper bounds on variances instead of expectation values.

Exactly as you said, but you don't need `hamiltonian_analysis` at all. You can just use `estimate_nmeas_for_frames` and not pass any expectation values.

[the remark this answers is not in the document]
Comment (max-radin) [reply steps/measure.py:160]: Yeah having `hamiltonian_analysis` would really just be a small convenience - it would remove the need to have a separate step in your workflow to group your qubit operator. (Or I guess you could write a step that wraps it in a list.)

**Review by max-radin — APPROVED** [review]

File: src/python/zquantum/core/hamiltonian.py [hunk 4]
```diff
@@ -103,6 +109,18 @@ def get_decomposition_function(
...
+def _calculate_variance_upper_bound(group: QubitOperator) -> float:
+    coefficients = np.array(list(group.terms.values()))
+    return np.sum(coefficients ** 2)
+
+
+def _remove_constant_term_from_group(group: QubitOperator) -> QubitOperator:
+    new_group = copy.deepcopy(group)
+    if new_group.terms.get(()):
+        del new_group.terms[()]
+    return new_group
```
Comment (jgonthier) [inline src/python/zquantum/core/hamiltonian.py:121]: What happens if the only term in the group is the constant term, as seems to be the case for `group_comeasurable_terms_greedy` above?

File: src/python/zquantum/core/hamiltonian.py [hunk 5]
```diff
@@ -201,49 +216,18 @@ def get_expectation_values_from_rdms(
...
     for term, coefficient in terms_iterator:
         reordered_qubitoperator += QubitOperator(term, coefficient)
-
     expectations_packed = interactionrdm.get_qubit_expectations(reordered_qubitoperator)
 
-    if () in expectations_packed.terms:
-        del expectations_packed.terms[
-            ()
-        ]  # Expectation of the constant term is excluded from expectation values
-
```
Comment (jgonthier) [inline src/python/zquantum/core/hamiltonian.py:211]: I see a potential issue here: now the expectation values returned by this function contain the constant term, while `compute_group_variance` expects:
- ExpectationValues without the constant term
- QubitOperator with the constant term (that gets removed)

File: tests/zquantum/core/hamiltonian_test.py [hunk 6]
```diff
@@ -303,6 +324,14 @@ def test_get_expectation_values_from_rdms_for_qubitoperator_list(
...
             np.array([2.0, 2.0]),
         ),
+        (
+            [
+                QubitOperator("[Z0 Z1] + [Z0] + []"),
+                QubitOperator("[X0 X1] + [X0] + 5 []"),
+            ],
+            None,
+            np.array([2.0, 2.0]),
+        ),
```
Comment (jgonthier) [inline tests/zquantum/core/hamiltonian_test.py:334]: I think we should add a case where one of the groups is just the constant term (related to my comment above)

**Review by jgonthier — CHANGES_REQUESTED** [review]
I think there is a potential issue for user error (again) in `compute_group_variances`, unless maybe the constant term expectation value is always at the end or something...

One fix could be to assume the QubitOperator and ExpectationValues given are consistent, and then eliminate both the constant term from the QubitOperator and the corresponding ExpectationValue.

File: src/python/zquantum/core/hamiltonian.py (hunk 4 above)
Comment (mstechly) [reply src/python/zquantum/core/hamiltonian.py:121]: It works (added a test for that in https://github.com/zapatacomputing/z-quantum-core/pull/268/commits/17e16d8ed08d4f29bb0f4e872990b354f685d1e0)

File: src/python/zquantum/core/hamiltonian.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aakunitsa) [reply src/python/zquantum/core/hamiltonian.py]: Thank you!

File: steps/measure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aakunitsa) [reply steps/measure.py:161]: @max-radin and @mstechly, sounds good! I am convinced now that `hamiltonian_analysis` is obsolete and should be removed! Thanks, guys!

**Review by jgonthier — APPROVED** [review]
After discussion, and with one additional test case, this is good to go! Thank you so much!
